temporal/error.go

```diff
@@ -166,6 +168,17 @@ var (
 	ErrSkipScheduleUpdate = internal.ErrSkipScheduleUpdate
 )
 
+// ApplicationErrorAttributes should be used to set all the desired attributes of a new ApplicationError
+// To get a new instance use ErrorAttributes function
+type ApplicationErrorAttributes = internal.ApplicationErrorAttributes
+
+// NewApplicationErrorWithExtraRequests creates new instance of *ApplicationError type, all the attributes of the
+// newly created error could be controlled through instance of ApplicationErrorAttributes.
+// This function also receives some extra requests. See activity.ActivityExtraRequests for details.
+func NewApplicationErrorWithExtraRequests(msg, errType string, attributes ApplicationErrorAttributes, extras ExtraRequests) error {
+	return internal.NewApplicationErrorWithExtraRequests(msg, errType, attributes, extras)
+}
+
```

Comment:
I thought per https://github.com/temporalio/sdk-go/pull/1319#discussion_r1425865743 we were going to do something like this:

```go
type ApplicationErrorOptions struct {
    Details []any
    Cause error
    NonRetryable bool
    NextAttemptRetryDelay time.Duration
}

func NewApplicationErrorWithOptions(message, errType string, options ApplicationErrorOptions) error
```

I strongly suggest we take this approach. (will save review of the other parts until after this fundamental part is decided)

Reply:
@cretz This is what I thought we discussed in December. I captured the screenshot for my notes as outcome. The comment you mentioned is regarding moving message and error type out of the error attribvutes structure which makes them required at the compile time. Please let me know if you would like to discuss further. 
<img width="1357" alt="Screenshot 2024-01-11 at 2 31 26 PM" src="https://github.com/temporalio/sdk-go/assets/149840466/4bf02239-fb8f-4c6b-86b5-16c6f8dedc18">

Reply:
I do think the single `WithOptions` and a single options set is simpler. Especially since now this will only be an error option, this is no different than any other optional field on application error I think.

Reply:
I think it's okay to go with the first option (single Options struct) and document that _currently_ the retry delay is only applicable in activities.